Design System: Add Context Menu

samwhale opened the pull request:
## Context

Building out the design system.

## Summary

Adds the `ContextMenu` component. [Designs](https://www.figma.com/file/bMhG3KyrJF8vIAODgmbeqT/Design-System?node-id=5733%3A75504)

|Hover|Keyboard|
|--|--|
|<image src="https://user-images.githubusercontent.com/22185279/107668982-8506db00-6c4e-11eb-96cf-22a6f628dd5b.gif" width="250px" />|<image src="https://user-images.githubusercontent.com/22185279/107668979-846e4480-6c4e-11eb-9158-1f592f0ac566.gif" width="250px" />|
|<image src="https://user-images.githubusercontent.com/22185279/107668975-833d1780-6c4e-11eb-953f-94cb4d351d6d.gif" width="250px" />|<image src="https://user-images.githubusercontent.com/22185279/107668971-80dabd80-6c4e-11eb-9e30-06aa140943af.gif" width="250px" />|

## Relevant Technical Choices

I decided not to stick to the old implementation. Instead, each item can have its own `onClick` event handler or `href` link.
- If `onClick` is passed on an item, a `<button>` is rendered
- If `href` is passed on an item, an `<a>` is rendered
- Links cannot be disabled. IMO that doesn't make sense. If the link should be 'disabled', we probably should not render it at all.
**Note**: This will be more difficult to integrate with the current popover menus since those are built off of one event handler, but should not be a nightmare when replacing. There are less props to deal with and we get the accessibility for free from `Button` and `Link`.

## To-do

- [x] Chatted with design to figure out hover/active styling

## User-facing changes

none

## Testing Instructions

- Design System -> ContextMenu
- Play with it and admire the beauty

### QA

<!--
Not all changes require manual QA.
-->

<!-- ignore-task-list-start -->
- [x] This is a non-user-facing change and requires no QA
<!-- ignore-task-list-end -->

### UAT

<!--
Sometimes the testing instructions for UAT can differ from the ones for QA.
-->

<!-- ignore-task-list-start -->
- [ ] UAT should use the same steps as above.
<!-- ignore-task-list-end -->

<!--
If the above checkbox has not been checked, write down all steps necessary for user acceptance testing take to test this PR.
-->
This PR can be tested by following these steps:

1. Visit storybook here: https://google.github.io/web-stories-wp/storybook/?path=/story/designsystem-components-contextMenu--default

## Reviews

### Does this PR have a security-related impact?

no

### Does this PR change what data or activity we track or use?

no

### Does this PR have a legal-related impact?

no

## Checklist

<!-- Check these after PR creation -->

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testiing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This PR contains automated tests (unit, integration, and/or e2e) to verify the code works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

<!--
Please reference the issue(s) this PR addresses.
No URLs, just the issue numbers.
Use "Fixes #123" if it fixes an issue.
Example:

Fixes #123
Partially addresses #456
See #789
-->

Fixes #6101

samwhale committed:
Initial context menu code

samwhale committed:
Small cleanup

samwhale committed:
Finish context menu v1

samwhale committed:
Add tests
* fix lint
* fix docs
* general cleanup

samwhale commented on assets/src/design-system/components/button/constants.js:
```diff
@@ -32,3 +32,5 @@ export const BUTTON_VARIANTS = {
   ICON: 'icon',
 };
+
+export const BUTTON_TRANSITION_TIMING = '0.6s ease 0s';
```
wanted to share this with the context menu so that the buttons animate at the same speed even though the styling is different

samwhale commented on assets/src/design-system/components/contextMenu/menu-item.js:
```diff
@@ -0,0 +1,120 @@
+    return <Link href={href}>{textContent}</Link>;
+  }
+
+  return <div>{textContent}</div>;
```
nice thing about rendering a div is that it doesn't focus with keyboard 😉

samwhale commented on assets/src/design-system/components/contextMenu/menu-item.js:
```diff
@@ -0,0 +1,120 @@
+    () => (
+      <>
+        <ItemText forwardedAs="span">{label}</ItemText>
+        {shortcut && <Shortcut forwardedAs="span">{shortcut}</Shortcut>}
```
I think for LTR we may want to switch these, no?

samwhale replied:
Added LTR styling:
<image src="https://user-images.githubusercontent.com/22185279/107672116-cd73c800-6c51-11eb-9b77-4145b5c0c74c.gif" width="300px" />

samwhale committed:
Add dark mode context menu

samwhale committed:
Fix styling

samwhale committed:
Fix disabled styles

samwhale committed:
Fix hover + active styles

samwhale committed:
Add LTR support

samwhale committed:
so lid

samwhale committed:
cleanup dependencies

embarks commented on assets/src/design-system/components/contextMenu/menu-item.js:
```diff
@@ -0,0 +1,153 @@
+import { Link } from '../typography/link';
+import { Text } from '../typography/text';
+
+const reverseCss = ({ isLTR }) => css`
```
nice

embarks reviewed (approved):
looks great in storybook

samwhale committed:
Add aria label to link

samwhale committed:
Fix hover and active styling for shortcut on link

BrittanyIRL commented on assets/src/design-system/components/contextMenu/index.js:
```diff
@@ -0,0 +1,162 @@
+    background-color: ${theme.colors.bg.primary};
+    border-radius: ${theme.borders.radius.small};
+    margin: 0;
+    min-width: 210px;
```
I'm seeing a width of 200px here: https://www.figma.com/file/bMhG3KyrJF8vIAODgmbeqT/Design-System?node-id=5733%3A75504 which is what i _think_ is the base 'lego' for the right click menus. 

![Screen Shot 2021-02-11 at 1 29 18 PM](https://user-images.githubusercontent.com/10720454/107695003-42a0c680-6c6d-11eb-88f6-6642a7c78f3b.png)

BrittanyIRL commented on assets/src/design-system/components/contextMenu/index.js:
```diff
@@ -0,0 +1,162 @@
+);
+
+const ContextMenu = ({ isOpen, isLTR, items, ...props }) => {
+  const ids = useMemo(() => items.map(() => uuidv4()), [items]);
```
i love this!

BrittanyIRL commented on assets/src/design-system/components/contextMenu/index.js:
```diff
@@ -0,0 +1,162 @@
+          <li
+            key={ids[index]}
+            className={
+              (separator === 'top' && 'separatorTop') ||
```
constants for these?

BrittanyIRL commented on assets/src/design-system/components/contextMenu/menu-item.js:
```diff
@@ -0,0 +1,153 @@
+import { Link } from '../typography/link';
+import { Text } from '../typography/text';
+
+const reverseCss = ({ isLTR }) => css`
```
I'm curious to see how this works in a wordpress setting where the elements get reversed? But I like it for the time being! When  the dashboard context menus  get swapped this will be fun to look at in wordpress.

samwhale replied:
Removed this since it's not necessary

BrittanyIRL commented on assets/src/design-system/components/contextMenu/menu-item.js:
```diff
@@ -0,0 +1,153 @@
+ * @return {Error|null} Returns an error if the conditions have not been met.
+ * Otherwise, returns null.
+ */
+export const linkOrButtonValidator = function (props, _, componentName) {
```
❤️

BrittanyIRL commented on assets/src/design-system/components/contextMenu/index.js:
```diff
@@ -0,0 +1,162 @@
+
+  return (
+    <Popover isOpen={isOpen} {...props}>
+      <MenuContainer>
```
I _feel like_ this could benefit from an aria label,  samesies with the list items since the spans inside aren't getting picked up in the ol' accessibility tree. 
<img width="767" alt="Screen Shot 2021-02-11 at 1 47 44 PM" src="https://user-images.githubusercontent.com/10720454/107696877-bba11d80-6c6f-11eb-9a56-3234b239fbc9.png">

samwhale replied:
Chatted with Brittany about this but posting here for visibility:

When adding a label to the list item the label gets repeated twice since screenreaders read the list item and then the button/link inside of it. We can't use `aria-hidden` on the inner content since they are focusable. 

Looked around online to see what the standard is here. Didn't find much, but it looks like these people don't label the list items either: http://web-accessibility.carnegiemuseums.org/code/navigation/

For now we'll leave it off

BrittanyIRL reviewed (changes requested):
This is gorgeous and I really love the updates you made to the structure of this. The one that exists in the dashboard right now has gone through many growing pains so this chance for a fresh start to refactor is great. 

For padding, let's stick to the 8pt grid so where it's 5px, let's do 4px and on contextMenu/index line 72 let's do 8px or 12px and 16px? Unless that just feels awful. I know Sam's said that there's sometimes a little spacing off and to try and stick to this rule if the designs are off by a little. 

The menu should also be able to be navigated with 'arrowDown' (sorry), since it _appears_ to be a list. 
![no arrow keys just tab](https://user-images.githubusercontent.com/10720454/107696384-1ede8000-6c6f-11eb-9b77-f1cb81233256.gif)

maxyinger commented on assets/src/design-system/components/contextMenu/index.js:
```diff
@@ -0,0 +1,162 @@
+  const ids = useMemo(() => items.map(() => uuidv4()), [items]);
+
+  return (
+    <Popover isOpen={isOpen} {...props}>
```
just learned about this but `role="dialog"` here could be good

maxyinger commented on assets/src/design-system/components/contextMenu/menu-item.js:
```diff
@@ -0,0 +1,153 @@
+import { Link } from '../typography/link';
+import { Text } from '../typography/text';
+
+const reverseCss = ({ isLTR }) => css`
```
Are these not aren't already handled by the `stylisRTLPlugin` ?

If you want to, you can just wrap your storybook instance in it and use the toggle to dictate it's application. will do the same thing it does in the editor:
```
import { StyleSheetManager, ThemeProvider } from 'styled-components';
import stylisRTLPlugin from 'stylis-plugin-rtl';
...
<StyleSheetManager stylisPlugins={isRTL ? [stylisRTLPlugin] : []}>
  // storybook component
</StyleSheetManager>
```

samwhale replied:
Removed my extra styling, it's not needed!

maxyinger commented on assets/src/design-system/components/contextMenu/menu-item.js:
```diff
@@ -0,0 +1,153 @@
+ * Otherwise, returns null.
+ */
+export const linkOrButtonValidator = function (props, _, componentName) {
+  if (props.onClick && props.href !== undefined) {
```
nice validator!
not following why `href` is being compared agains `undefined` and `onClick` is not? same with the `href` & `disabled` use below.

samwhale replied:
yeah good callout. 

`href`
The reason I'm checking href against `undefined` is because if `href=""` then `!props.href` will be `true`. This just ensures that if it's an empty string it will still count it.

`disabled`
If a person passes in an `href` but says `disabled={false}` I still wanted the proptypes to yell at the person since it shouldn't be specified at all. 

Thoughts?

maxyinger commented on assets/src/design-system/components/contextMenu/stories/index.js:
```diff
@@ -0,0 +1,141 @@
+  { label: 'Clear text styles' },
+  { label: 'Add style to "Saved styles"' },
+  { label: 'Add color to "Saved colors' },
+];
```
Would it be better just to require an `id` on each item in the data getting passed in vs generating one at run time? probably doesn't matter all too much  either way perf wise cus the lists shouldn't be changing frequently or have large subtrees of children.

samwhale replied:
Toatlly - I'm not sure which I prefer. I think for initial usage I like not having `id` be a field that has to be defined by the dev. Less to deal with 🤷 

I like both options, but maybe we keep it like this for now?

maxyinger commented on assets/src/design-system/components/contextMenu/stories/index.js:
```diff
@@ -0,0 +1,141 @@
+      </Container>
+    </DarkThemeProvider>
+  );
+};
```
You might just be able to do this to see how much custom rtl support you actually need to add. It does a surprising amount automagically

```suggestion
import { StyleSheetManager, ThemeProvider } from 'styled-components';
import stylisRTLPlugin from 'stylis-plugin-rtl';

export const DarkMode = () => {
  const itemsWithEventHandlers = items.map((item) => ({
    ...item,
    onClick: action(`Clicked on \`${item.label}\``),
  }));

  return (
    <DarkThemeProvider>
      <StyleSheetManager stylisPlugins={boolean('Right to Left', false) ? [stylisRTLPlugin] : []}>
        <Container>
          <ContextMenu
            items={itemsWithEventHandlers}
            isOpen={boolean('isOpen', true)}
          />
        </Container>
      </StyleSheetManager>
    </DarkThemeProvider>
  );
};
```

samwhale replied:
This works great in the actual app (when I removed the custom `isLTR` styling). 

I think storybook has some challenges when I do it there:

|Storybook|Dashboard|
|--|--|
|<image src="https://user-images.githubusercontent.com/22185279/107705673-556ec780-6c7c-11eb-9b90-d88978dfd304.png" width="200px" />|<image src="https://user-images.githubusercontent.com/22185279/107705625-41c36100-6c7c-11eb-80e1-8ece6414dd48.png" width="200px" />|

swissspidy replied:
That's because the app has a "global" `StyleSheetManager` wrapper around it, which we don't have in the editor.

We could look into adding that for all storybook stories as well. For the short term this suggestion seems to work.

maxyinger reviewed (approved):
Looks  👏  great 👏   works  👏  great 👏

samwhale committed:
Pr Fixes
- remove isLTR, we don't need it
- add constants for classnames
- add aria label
- fix width and font size of menu

samwhale committed:
Align spacing to multiples of `4`

samwhale committed:
Add keyboard navigation of list with arrow keys

samwhale committed:
Small cleanup for new function

samwhale committed:
Remove id

samwhale commented on assets/src/design-system/components/contextMenu/index.js:
```diff
@@ -0,0 +1,228 @@
+                return;
+              }
+              index--;
+            }
```
a little bit of repeated code here but the direction (ascending/descending) is different. I could abstract this into a function and to pick a direction but I think it would make it less readable.

Thoughts?

samwhale replied:
To be fair: it's already kind of hard to read.

This looks for the next focusable element ('a' or 'button'). If none are found, then it keeps the focus on the current element.

BrittanyIRL replied:
We have a util actually to help with keyboard events. I know the current implementation has a switch like this, but the editor has a lot of focus traps and it'd probably be good to standardize how keyboard handlers are done? 

Something like this would clean it up - this works as is for nav, just not disabled items.
```
  const totalIndex = useMemo(() => items.length - 1, [items]);

  const handleKeyboardNav = useCallback(
    ({ key }) => {
      if ([isRTL ? 'ArrowDown' : 'ArrowUp'].includes(key)) {
        setFocusedIndex((currentIndex) =>
          currentIndex - 1 > -1 ? currentIndex - 1 : -1
        );
      }

      if ([isRTL ? 'ArrowUp' : 'ArrowDown'].includes(key)) {
        setFocusedIndex((currentIndex) =>
          currentIndex + 1 <= totalIndex ? currentIndex + 1 : currentIndex
        );
      }
    },
    [isRTL, totalIndex]
  );
  useKeyDownEffect(listRef, { key: ['down', 'up'] }, handleKeyboardNav, [
    handleKeyboardNav,
  ]);
```

samwhale replied:
Chatted with brittany about this - we don't need to worry about RTL/LTR since it's not flipping vertically

samwhale commented on assets/src/design-system/components/contextMenu/menu-item.js:
```diff
@@ -0,0 +1,167 @@
+
+  const handleFocus = () => {
+    setFocusedIndex(index);
+  };
```
Keeps focusability interchanging `tab` and arrow keys.

samwhale committed:
Reduce re-renders

samwhale commented on assets/src/design-system/components/contextMenu/menu-item.js:
```diff
@@ -0,0 +1,169 @@
+}) => {
+  const itemRef = useRef(null);
+
+  const handleFocus = () => {
+    if (focusedIndex !== index) {
+      setFocusedIndex(index);
+    }
+  };
```
Keeps focus functionality consistent even if we are interchanging between `tab` and arrow keys.

samwhale committed:
Refactor handleKeyDown

BrittanyIRL commented on assets/src/design-system/components/contextMenu/index.js:
```diff
@@ -0,0 +1,212 @@
+    background-color: ${theme.colors.bg.primary};
+    border-radius: ${theme.borders.radius.small};
+    margin: 0;
+    /* min-width: 200px; */
```
nit: remove if unnecessary?

BrittanyIRL commented on assets/src/design-system/components/contextMenu/index.js:
```diff
@@ -0,0 +1,212 @@
+   */
+  const handleKeyboardNav = useCallback(
+    ({ key }) => {
+      const isAscending = key === KEYS.UP;
+      let index = focusedIndex + (isAscending ? -1 : 1);
+      let terminate = isAscending ? index < 0 : index > totalIndex;
```
👍

BrittanyIRL reviewed (approved):
Awesome! 

Looks like there's some circular structure errors in storybook if you click a menu item twice but otherwise this works awesome! Excited to have a new version of this. 
<img width="588" alt="Screen Shot 2021-02-12 at 2 22 02 PM" src="https://user-images.githubusercontent.com/10720454/107823913-d17d1400-6d3d-11eb-8865-65195db99887.png">

samwhale committed:
Remove comment, fix circular import

samwhale committed:
Change imports
